README.md
# audio_wave_interpreter

Project for ECE243 - Computer Organization in University Of Toronto

Implemented in NIOSII assembly code.
Board Used: DE1-SOC

Program samples microphone input and either displays the frequency spectrum, or the sound wave.


Changelog
Changelog for Carson and Carrie: (Jiachen Liu and Kaixuan Wang, Team 60)

March 29, 2018
-	Carson: Research on Fast Fourier Transform
-	Carrie: Research on Fast Fourier Transform
March 30, 2018
-	Carson: Reimplementation of wave drawing to account for refresh rate of monitor
-	Carrie: Tried different Fast Fourier Transform algorithm 
March 31, 2018
-	Carson: hanged wave drawing to sample and store into array then print from array
-	Carrie: Implemented Fast Fourier Transform c algorithm and tested it
April 1, 2018
-	Carson: Added refresh rate wait
-	Carrie: Linked algorithm to assembly project, it takes double array input and returns a double array of amplitude (frequency domain)

Aprril 5, 2018
-	Carson: working on normalizing audio wave & frequency using switches
-	Carrie: Fast fourier transform bugfixing(problem with data size)

Aprril 6, 2018
-	Carson: wrote code for displaying lines on vga, which can be used for displaying freq
-	Carrie: linked c file to assembly, working on connect it to fifo 

Aprril 7, 2018
-	Carson: wrote code for keyboard interrupt (reads user input and decide which to show wave/frequency)
-	Carrie: working on extracting data from fifo and return data array to proper address

Aprril 8, 2018
-	Carson: linked all parts together
-	Carrie: fixing displaying frequency 
Aprril 9, 2018
-	Carson: final editing
-	Carrie: final editing

Test_fft.s
.global main
main:

movia r10, 0x03FFFFFF
subi sp, r10, 68
#float representation of 1
movia r2, 0x3F800000

stw r2, 0(r10)
stw r0, 4(r10)
stw r0, 8(r10)
stw r0, 12(r10)
stw r2, 16(r10)
stw r0, 20(r10)
stw r0, 24(r10)
stw r0, 28(r10)

mov r4, r10
call fft_func();

Loop:
	br Loop


fft.c

#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#define SIZE 256
#define SWAP(a,b) tempr=(a); (a)=(b); (b)=tempr


/*
* Implementation of FFT done by http://www.guitarscience.net/papers/fftalg.pdf
* Simple Fast Fourier Transformation Algorithms in C
*
* Has been modified to work better with the assembly code.
*/

int* fft_func(int* data_input) {
    float wtemp, wr, wpr, wpi, wi, theta;
    float tempr, tempi;
    int N = SIZE;
    float data1[2 * SIZE];
    int i = 0, j = 0, n = 0, k = 0, m = 0, isign = -1, istep, mmax;
    printf("\nSize of sample: %d", N);

    for (unsigned i = 0; i<N * 2; i = i + 2) {
        data1[i] = 0;
        data1[i + 1] = data_input[i / 2];
    }

    float *data;
    data = &data1[0] - 1;
    n = N * 2;
    j = 1;

    // do the bit-reversal
    for (i = 1; i < n; i += 2) {
        if (j > i) {
            SWAP(data[j], data[i]);
            SWAP(data[j + 1], data[i + 1]);
        }
        m = n >> 1;
        while (m >= 2 && j > m) {
            j -= m;
            m >>= 1;
        }
        j += m;
    }
    // calculate the FFT
    mmax = 2;
    while (n > mmax) {
        istep = mmax << 1;
        theta = isign * (6.28318530717959 / mmax);
        wtemp = sin(0.5 * theta);
        wpr = -2.0 * wtemp*wtemp;
        wpi = sin(theta);
        wr = 1.0;
        wi = 0.0;
        for (m = 1; m < mmax; m += 2) {
            for (i = m; i <= n; i += istep) {
                j = i + mmax;
                tempr = wr * data[j] - wi * data[j + 1];
                tempi = wr * data[j + 1] + wi * data[j];
                data[j] = data[i] - tempr;
                data[j + 1] = data[i + 1] - tempi;
                data[i] = data[i] + tempr;
                data[i + 1] = data[i + 1] + tempi;
            }

            wtemp = wr;
            wr += wtemp * wpr - wi * wpi;
            wi += wtemp * wpi + wi * wpr;
        }
        mmax = istep;
    }

    unsigned x = 0;
    for (k = 0; k < 2 * N; k += 2) {
        int temp = round(sqrt(data[k + 1] * data[k + 1] + data[k + 2] * data[k + 2]));
        data_input[x]=temp;
        x = x + 1;
    }

    return data_input;
} // end of dittt()



fft_display.s
.equ WIDTH, 320
.equ HEIGHT, 240
.equ ADDR_AUDIODACFIFO, 0xFF203040
.equ ADDR_SLIDESWITCHES, 0xFF200040
.equ TIMER1, 0xFF202000
.equ AMP_START, 200
.equ SIZE, 2048

.data
DATA_IN:
    .skip 2052

.text

/*
 * Function for drawing the frequency spectrum.
 * Actual Fast Fourier Transform is being done by a C program.
 * This function stores samples into an array and 
 * passes it to the C program
 * If size is to be changed, both in this file as well as the 
 * C program will have to be changed.
 * No parameters or return values.
 */

.global drawFreq
drawFreq:
    subi sp, sp, 16
    stw ra, 0(sp)
    stw r17, 4(sp)
    stw r18, 8(sp)
    stw r21, 12(sp)

    /* Determine amount to attentuate signal by */
    movia r2, ADDR_SLIDESWITCHES
    movi r17, 1
    ldwio r2, 0(r2) #determine shift amount
    sll r17, r17, r2 # 2^(r2) 

    /* Getting data pointer to array*/
    movia r21, DATA_IN
    subi r21, r21, 2052 #start on a lower address rather than larger
    mov r10, r0

    /*
     * We clear the fifo, so that we can start with fresh data,
     * rather than old data which we are unsure about the 
     * ordering of
     */

    clear_fifo: 
    movia r18, ADDR_AUDIODACFIFO
    ldwio r2,12(r18) # Load mic
    ldwio r2,8(r18)
    ldwio r2, 4(r18)      # Read fifospace register 
    andi  r2, r2, 0xff    # Extract # of samples in Input Right Channel FIFO 
    bne   r2, r0, clear_fifo  # If no samples in FIFO, go back to start 
    
    /*
     * Loading data into the array
     * signal is being normalized
     */
    
    check_fifo: 
    movia r18, ADDR_AUDIODACFIFO
    ldwio r2, 4(r18)      # Read fifospace register 
    andi  r2, r2, 0xff    # Extract # of samples in Input Right Channel FIFO 
    beq   r2, r0, check_fifo  # If no samples in FIFO, go back to start 
    
    ldwio r2,12(r18) # Load mic
    ldwio r2,8(r18) 
    
    subi r21, r21, 4 # move to next array location
    div r2, r2, r17 # 2^(r2) normalization
    stw r2, 0(r21) # Storing sample to stack

    addi r10, r10, 1 # Increase counter
    movia r18, WIDTH-1 # Check if enough samples

    beq r10, r18, fft #draw screen if samples are enough
    br check_fifo
    

fft:
    mov r4 , r21 #pointer to array
    call fft_func()

    mov r21, r2 
    movia r10, WIDTH-32

    draw_freq:
    subi r10, r10, 1
    addi r21, r21, 4

    mov r4, r0
    mov r5, r10
    call FillLine

    movia r4, 0xffffffff
    mov r5, r10
    ldw r6, 0(r21)
    call FillSegment

    subi r10, r10, 1

    mov r4, r0
    mov r5, r10
    call FillLine

    movia r4, 0xffffffff
    mov r5, r10
    ldw r6, 0(r21)
    call FillSegment

    movi r2, 32
    bne r10, r2, draw_freq

    /*
     * Control for the refresh rate of our wave drawing
     */
     
    movia r18, TIMER1 #timer initialization
    movia r2, 0xB9AA #60 Hz
    stwio r2, 8(r18)
    movia r2, 0x65
    stwio r2, 12(r18)
    movi r2, 0x4
    stwio r2, 4(r18) #timer start
    
    check_timer: ldwio r2, 0(r18)
    andi r2, r2, 0x01
    beq r2, r0, check_timer
    stwio r0, 0(r18) #clearing timer
    br freq_exit

freq_exit:
    ldw ra, 0(sp)
    ldw r17, 4(sp)
    ldw r18, 8(sp)
    ldw r21, 12(sp)
    addi sp, sp, 16
    ret

main_keyboard.s
.equ ADDR_PS2, 0xFF200100
.equ STACK_POINTER, 0x04000000
.equ SCANCODE_w, 0x1D
.equ SCANCODE_f, 0x2B
.equ FREQ_TOGGLE, 0x01
.equ IRQ_7, 0x080

.text

/*
 * Main program loop to keep printing the wave or frequency.
 * When a keyboard interrupt is detected, the VGA screen will toggle between 
 * drawing the sound wave and the frequency spectrum
 */

.global main
main:

    setup: movia sp, STACK_POINTER    # Initial stack pointer 

    /*
     * Initialization of the screen with black
     */

    mov r4, r0 #filling the screen with black
    call FillColour
    
    /*
     * Enable interrupt for the PS/2 Keyboard
     */

    movia r18, ADDR_PS2 
    movia r2, 0x80 #IRQ line 7
    wrctl ienable, r2
    movia r2, 0x01 #PIE:0 
    stwio r2, 4(r18)
    wrctl status, r2
    movia r16, FREQ_TOGGLE 
    
    /*
     * Main loop for detecting keyboard input 
     * and calling proper drawing function 
     * interrupt can happen inside of the drawing functions
     */    

    MAIN_LOOP:
    movia r17, SCANCODE_w
    movia r18, SCANCODE_f 
    
    beq et, r17, set_wave
    beq et, r18, set_freq
    
    display_branch:
    bne r16, r0, 2f
        1:
        subi sp, sp, 4
        stw r10, 0(sp)
        call drawFreq
        ldw r10, 0(sp)
        addi sp, sp, 4 
        br MAIN_LOOP
        
        2:
        subi sp, sp, 4
        stw r10, 0(sp)
        call drawWave
        ldw r10, 0(sp)
        addi sp, sp, 4
        br MAIN_LOOP
    
/*
 * Branches for setting the 
 * wave or display toggle
 */

set_wave:
    movi r16, FREQ_TOGGLE
    br display_branch
    
set_freq:
    mov r16, r0
    br display_branch
    

.section .exceptions, "ax"

HANDLER:
    subi sp, sp, 28
    stw r2, 0(sp)
    stw r18, 4(sp)
    stw r4, 8(sp)
    stw r5, 12(sp)
    stw r6, 16(sp)
    stw r2, 20(sp)
    stw r3, 24(sp)

    /*
     * Checking if IRQ line 7 is the one that is 
     * activated, keyboard interrupt
     */

    rdctl r2, ipending
	movi r3, 0x7
    srl r2, r2, r3
    andi r2, r2, 0x01
    beq r2, r0, interrupt_return
	
    movia r18, ADDR_PS2
    
    check_ps2_fifo:
    ldbio et, 0(r18) #load the character
    ldwio r2, 0(r18) #check if anymore valid
    srli r2, r2, 0x0f #checking valid bit 15
    andi r2, r2, 1 
    bne r2, r0, check_ps2_fifo
    
    interrupt_return:
    ldw r2, 0(sp)
    ldw r18, 4(sp)
    ldw r4, 8(sp)
    ldw r5, 12(sp)
    ldw r6, 16(sp)
    ldw r2, 20(sp)
    ldw r3, 24(sp)
    addi sp, sp, 28
    subi ea, ea, 4
    eret

vga_functions.s
.equ WIDTH, 319
.equ HEIGHT, 239
.equ X_OFFSET, 10
.equ Y_OFFSET, 1
.equ PIXELBUFFER, 0x08000000 # Pixel buffer. Same on all boards.
.equ AMP_START, 210

.text
.global FillLine
.global FillColour
.global WritePixel
.global FillSegment

FillSegment:
    subi sp, sp, 20
    stw r15, 0(sp)
    stw r16, 4(sp)
    stw r17, 8(sp)
    stw r18, 12(sp)
    stw ra, 16(sp)
    
    mov r18, r4
    mov r16, r5
    movi r2, AMP_START #get row index for height
    sub r15, r2, r6
    bge r15, r0, start_segment #check if number is larger
        mov r15, r0
        
    # Two loops to draw each pixel
    start_segment:  movi r17, AMP_START
        fs_1: subi r17, r17, 1  
            mov r4, r16
            mov r5, r17
            mov r6, r18
            call WritePixel   
            bgt r17, r15, fs_1
    
    ldw r15, 0(sp) 
    ldw r16, 4(sp)
    ldw r17, 8(sp)
    ldw r18, 12(sp)
    ldw ra, 16(sp)
    addi sp, sp, 20
    ret


# r4: colour
FillColour:
    subi sp, sp, 16
    stw r16, 0(sp)      # Save some registers
    stw r17, 4(sp)
    stw r18, 8(sp)
    stw ra, 12(sp)
    
    mov r18, r4
    
    # Two loops to draw each pixel
    movi r16, HEIGHT
    fc_1: movi r17, WIDTH
        subi r16, r16, 1
        fc_2:  subi r17, r17, 1
            mov r4, r17
            mov r5, r16
            mov r6, r18
            call WritePixel     # Draw one pixel
            bgt r17, r0, fc_2
        bgt r16, r0, fc_1
    
    ldw ra, 12(sp)
    ldw r18, 8(sp)
    ldw r17, 4(sp)
    ldw r16, 0(sp)    
    addi sp, sp, 16
    ret
    
    
# r4: colour
# r5: col

FillLine:
    subi sp, sp, 16
    stw r16, 0(sp)      # Save some registers
    stw r17, 4(sp)
    stw r18, 8(sp)
    stw ra, 12(sp)
    
    # Two loops to draw each pixel
    movi r17, HEIGHT+1
    mov r18, r4
    mov r16, r5
        fl_1:  subi r17, r17, 1
            mov r4, r16
            mov r5, r17
            mov r6, r18
            call WritePixel
            bgt r17, r0, fl_1
    
    ldw ra, 12(sp)
    ldw r18, 8(sp)
    ldw r17, 4(sp)
    ldw r16, 0(sp)    
    addi sp, sp, 16
    ret

# r4: col (x)
# r5: row (y)
# r6: colour value
WritePixel:   
    slli r5, r5, X_OFFSET #make space for col offset
    slli r4, r4, Y_OFFSET #offset
    add r5, r5, r4 #add offset together
    movia r4, PIXELBUFFER 
    add r5, r5, r4 #offset the pixel buffer
    sthio r6, 0(r5)
    ret


waveform_display.s
.equ WIDTH, 320
.equ HEIGHT, 240
.equ ADDR_AUDIODACFIFO, 0xFF203040
.equ ADDR_SLIDESWITCHES, 0xFF200040
.equ TIMER1, 0xFF202000
.equ PIXBUF, 0x08000000	# Pixel buffer. Same on all boards.
.equ MAX_ROW, 239
.equ MIN_ROW, 0
.equ CENTER_ROW, 119

.data
WAVE_DATA:
    .skip 1288

.text

/*
 * Function which draws a sound wave from the microphone input onto the VGA.
 * No values need to be passes, there is also no return values. Values that are
 * too big or small will be normalized to the max and min of the screen
 */

.global drawWave
drawWave:
    subi sp, sp, 16
    stw ra, 0(sp)
    stw r17, 4(sp)
    stw r18, 8(sp)
    stw r21, 12(sp)

    movia r21, WAVE_DATA
    subi r21, r21, 1288

    mov r10 , r0 #sample counter
    
    /*
     * Reading the switches to determine the shift amount
     */

    shift_read: movia r2, ADDR_SLIDESWITCHES
	movi r17, 1
    ldwio r2, 0(r2) #determine shift amount
	sll r17, r17, r2 # 2^(r2) 
    
    /*
     * clearing the fifo so that we always start at a clean slate
     */
     
    clear_fifo: movia r18, ADDR_AUDIODACFIFO
    ldwio r2,12(r18) # Load mic
    ldwio r2,8(r18)
    ldwio r2, 4(r18)      # Read fifospace register 
    andi  r2, r2, 0xff    # Extract # of samples in Input Right Channel FIFO 
    bne   r2, r0, clear_fifo  # If no samples in FIFO, go back to start 
    
    /*
     * extracting samples from the fifo
     */
    
    check_fifo: movia r18, ADDR_AUDIODACFIFO
    ldwio r2, 4(r18)      # Read fifospace register 
    andi  r2, r2, 0xff    # Extract # of samples in Input Right Channel FIFO 
    beq   r2, r0, check_fifo  # If no samples in FIFO, go back to start 
    
    ldwio r2,12(r18) # Load mic
    ldwio r2,8(r18) 
    div r18, r2, r17 # Normalize right shift by 2^(r2)
    
    subi r21, r21, 4
    stw r18, 0(r21) # Storing sample to stack
    addi r10, r10, 1 # Increase counter
    
    movi r18, WIDTH # Check if enough samples
    beq r10, r18, draw_screen #draw screen is samples are enough
   	br check_fifo
    
    
/*
 * draws the screen by going through the array and printing array values to the screen
 */
 
draw_screen:

	mov r10, r21
    movi r18, WIDTH-33
    
	draw_wave_start:
    mov r5, r18 #col
	mov r4, r0 #black
	call FillLine #erases previous
    
    mov r4, r18 #col
	movia r6, 0xffffffff #white
    
    ldw r17, 0(r10) #getting array value
    addi r5, r17, CENTER_ROW #centering at 120
    movi r2, 239 #lower limit
	bgt r5, r2, normalize_down #normalization, incase amplitude is too large
	blt r5, r0, normalize_up
    draw_amp_pixel: call WritePixel
    addi r10, r10, 4 #increase array value
    subi r18, r18, 1 #increase col value
    movi r2, 31
   	bne r18, r2, draw_wave_start
    
    /*
     * Control for the refresh rate of our wave drawing
     */
     
    movia r18, TIMER1 #timer initialization
    movia r2, 0xB9AA #60 Hz
    stwio r2, 8(r18)
    movi r2, 0x65
    stwio r2, 12(r18)
    movi r2, 0x4
    stwio r2, 4(r18) #timer start
    
    check_wave_timer: ldwio r2, 0(r18)
    andi r2, r2, 0x01
    beq r2, r0, check_wave_timer
    stwio r0, 0(r18) #clearing timer
    
    mov r10, r0 #reset counter
	movia r21, WAVE_DATA # Initial stack pointer
    br exit_drawWave
    
normalize_down:
	movi r5, 239
	br draw_amp_pixel

normalize_up:
	movi r5, 0
	br draw_amp_pixel

exit_drawWave:
    ldw ra, 0(sp)
    ldw r17, 4(sp)
    ldw r18, 8(sp)
    ldw r21, 12(sp)
    addi sp, sp, 16
    ret
